### ISSUE TITLE
[question] Why? input shape of softmax in HashingMemoryProduct 

### ISSUE THREAD

**Issue by richardyy1188** [facebookresearch/XLM#232]
In the code we used a multi-head mechanism, and every reading head will find their best knn keys that match it, and the scores. `(bs, heads, knn)`
but then apply softmax across heads by view scores as `(bs * heads, knn)`
```
# in get_indices()
s = torch.cat([s.view(bs, 1, self.knn) for s, _ in outputs], 1)  # (bs,heads,knn)  
i = torch.cat([i.view(bs, 1, self.knn) for _, i in outputs], 1)  # (bs,heads,knn)  
return s.view(-1, self.knn), i.view(-1, self.knn)
  
# in forward()
scores = F.softmax(scores.float(), dim=-1).type_as(scores)   # (bs*heads,knn)
```
  
Shouldn't it apply softmax on scores with size `(bs, heads, knn)` to avoid normalize scores that indicate similarity to different heads ?

**Comment by glample** [comment]
Hi,

So assuming that the scores are of shape `(bs * heads, knn)` (i.e. the way they come out of `get_indices`)

Then:
`scores = F.softmax(scores, dim=-1).view(bs, heads, knn)`
and:
`scores = F.softmax(scores.view(bs, heads, knn), dim=-1)`

will return identical outputs. So I don't think there is any issue here. The problem you mention would occur with:
`scores = F.softmax(scores.view(bs, heads * knn), dim=-1).view(bs, heads, knn)`
